deps/rabbit/src/rabbit_deprecated_features.erl

```diff
@@ -0,0 +1,480 @@
+      Default :: boolean(),
+      IsPermitted :: boolean().
+%% @private
+
+is_permitted_in_configuration(FeatureName, Default) ->
+    Settings = application:get_env(rabbit, permit_deprecated_features, #{}),
+    case maps:get(FeatureName, Settings, undefined) of
+        undefined ->
+            Default;
+        Default ->
+            PermittedStr = case Default of
+                               true  -> "permitted";
+                               false -> "not permitted"
+                           end,
+            ?LOG_DEBUG(
+               "Deprecated features: `~ts`: ~ts in configuration, same as "
+               "default",
+               [FeatureName, PermittedStr],
+               #{domain => ?RMQLOG_DOMAIN_FEAT_FLAGS}),
+            Default;
+        Permitted ->
+            PermittedStr = case Permitted of
+                               true  -> "permitted";
+                               false -> "not permitted"
+                           end,
+            ?LOG_DEBUG(
+               "Deprecated features: `~ts`: ~ts in configuration, overrides "
+               "default",
+               [FeatureName, PermittedStr],
+               #{domain => ?RMQLOG_DOMAIN_FEAT_FLAGS}),
+            ?assert(is_boolean(Permitted)),
+            Permitted
+    end.
+
+-spec maybe_log_warning(FeatureName, Permitted) -> ok when
+      FeatureName :: rabbit_feature_flags:feature_name(),
+      Permitted :: boolean().
+%% @private
+
+maybe_log_warning(FeatureName, Permitted) ->
+    case should_log_warning(FeatureName) of
+        false ->
+            ok;
+        true ->
+            Warning = get_warning(FeatureName),
+            FormatStr = "Deprecated features: `~ts`: ~ts",
+            FormatArgs = [FeatureName, Warning],
+            case Permitted of
+                true ->
+                    ?LOG_WARNING(
+                       FormatStr, FormatArgs,
+                       #{domain => ?RMQLOG_DOMAIN_FEAT_FLAGS});
+                false ->
+                    ?LOG_ERROR(
+                       FormatStr, FormatArgs,
+                       #{domain => ?RMQLOG_DOMAIN_FEAT_FLAGS})
+            end
+    end.
+
+-define(PT_DEPRECATION_WARNING_TS(FeatureName), {?MODULE, FeatureName}).
+
+-spec should_log_warning(FeatureName) -> ShouldLog when
+      FeatureName :: rabbit_feature_flags:feature_name(),
+      ShouldLog :: boolean().
+%% @private
+
+should_log_warning(FeatureName) ->
+    Key = ?PT_DEPRECATION_WARNING_TS(FeatureName),
+    Now = erlang:timestamp(),
+    try
+        Last = persistent_term:get(Key),
+        Diff = timer:now_diff(Now, Last),
+        if
+            Diff >= 24 * 60 * 60 * 1000 * 1000 ->
+                persistent_term:put(Key, Now),
+                true;
+            true ->
+                false
+        end
+    catch
+        error:badarg ->
+            persistent_term:put(Key, Now),
+            true
+    end.
+
+enable_underlying_feature_flag_cb(
```

Comment:
When the node is starting, is this executing only locally? When in `removed` phase, locally there will be no callback code at all, but somebody can still try upgrading from `denied_by_default` release straight to `removed`.

Does it make sense to limit the number of different phases that a cluster have for a given feature to 2 nearby phases (in `rabbit_ff_controller:merge_feature_flags/2`)?  So that it would be outright impossible to upgrade to a release that skips a phase. (Such jumps during upgrade are not supported anyway, but making this a hard limitation can allow us to get away only with a local check).

It's just that in the `disconnected` state the callback should still be there and actually called, and should be removed only at the same time as proceeding to `removed`. But if there is an absolute certainty that all other nodes already sit at least at `disconnected`, this absence of callback should no longer matter.

Reply:
Callbacks are executed locally, yes. And indeed, there will be no callback executed for `disconnected` or `removed` deprecated features. Those will be treated exactly like `required` feature flags.

It means a node must have the corresponding feature flags enabled (the deprecated feature are denied) in order to start or join a cluster. With required feature flags, if a user didn't enabled them before upgrading to the version requiring them, the node will fail to start. He will have to downgrade, enable the feature flags, upgrade again. This will be the same with disconnected and removed deprecated features.

Preventing an upgrade skipping a phase won't help because there is always a switch from non-required to required; here between `denied_by_default` and `disconnected` (or `removed` if `disconnected` is skipped).

Reply:
So in the end deprecated feature flags will be also handled by `rabbit_ff_controller:check_node_compatibility/1` (but this time as almost real features)? And there it will take into account whether deprecated feature is denied or not for all nodes.

Reply:
I should probably clarify: once we remove the compatibility code behind a feature flag, there is nothing left in RabbitMQ to prevent an upgrade. This action must come from something outside of RabbitMQ.

For instance, we made the `quorum_queue` feature flags required in 3.11.0 IIRC. Therefore, if the user upgraded from the very latest 3.10.x to 3.11.0 without enabling the feature flag first, the upgrade would proceed and the node would refuse to start.

A good solution would be the pre-install script in the package that we already mentionned. I don't know what could be done for image-based deployment.

That's why I'm saying that for disconnected/removed deprecated features, the problem would be the same.

Reply:
Another clarification: a disconnected deprecated feature would be behave exactly like a removed one. The only difference if we didn't remove the code from Git.

The idea is that for users who really really want to upgrade but continue to use the disconnected deprecated feature at the same time, then they are ok to recompile RabbitMQ themselves (after patching the deprecated feature definition to change the deprecation phase for instance).

We may not want to have that "disconnected" phase for all feature flags. For instance, we will probably not use it for classic mirrored queues: it will go from denied by default to removed in RabbitMQ 4.0 because we want to remove the code to finish with Khepri.